Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/TypeResolution/InteractableTypeFinder.cs

```diff
@@ -0,0 +1,140 @@
+        /// that type.
+        /// </summary>
+        private static Dictionary<Type, List<InteractableType>> cache = new Dictionary<Type, List<InteractableType>>();
+
+        /// <summary>
+        /// Gets the list of InteractableType objects for classes that support the specified types.
+        /// </summary>
+        private static List<InteractableType> FindTypes(List<Type> types, TypeRestriction typeRestriction)
+        {
+            EnsureCacheForTypes(types, typeRestriction);
+            return GetTypesFromCache(types);
+        }
+
+        /// <summary>
+        /// Gets the list of InteractableType objects for classes that support the specified types by
+        /// looking directly in the cache.
+        /// </summary>
+        /// <remarks>
+        /// Assumes it is called after EnsureCacheForTypes. Otherwise, this is dangerous to call.
+        /// </remarks>
+        private static List<InteractableType> GetTypesFromCache(List<Type> types)
+        {
+            List<InteractableType> interactableTypes = new List<InteractableType>();
+            foreach (Type type in types)
+            {
+                interactableTypes.AddRange(cache[type]);
+            }
+            return interactableTypes;
+        }
+
+        /// <summary>
+        /// Ensures a cache entry is setup for all types in the InteractableType enum.
+        /// </summary>
+        /// <remarks>
+        /// Note that this is not invoked at runtime and is assumed to be invoked from a single
+        /// threaded UI context, and is thus not locked.
+        /// </remarks>
+        private static void EnsureCacheForTypes(List<Type> types, TypeRestriction typeRestriction)
+        {
+            List<Type> cacheMisses = new List<Type>();
+            foreach (Type type in types)
+            {
+                if (!cache.ContainsKey(type))
+                {
+                    cacheMisses.Add(type);
+                }
+            }
+
+            var assemblies = AppDomain.CurrentDomain.GetAssemblies();
+            foreach (Type type in cacheMisses)
+            {
+                cache[type] = GetTypesFromAssemblies(type, typeRestriction, assemblies);
+            }
+        }
+
+        /// <summary>
+        /// Loads the classes that derive from the given type by looking through all of the assemblies.
+        /// </summary>
+        private static List<InteractableType> GetTypesFromAssemblies(Type type, TypeRestriction typeRestriction, Assembly[] assemblies)
+        {
+            List<InteractableType> interactableTypes = new List<InteractableType>();
+            foreach (Assembly assembly in assemblies)
+            {
+                foreach (Type assemblyType in assembly.GetTypes())
+                {
+                    TypeInfo info = assemblyType.GetTypeInfo();
+                    bool exactBaseMatch = typeRestriction == TypeRestriction.AllowBase && assemblyType.Equals(type);
```

Comment:
I may be misreading this, but isn't this just adding the Type you are searching for to the list; and since Types are unique there will be just this one added, so this check could be omitted.

Reply:
Yeah, you're reading this correctly. This was mostly a translation from the original code, which handled both the base case and the derived case in the same if statement. I just moved this out of this block.